game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/fire/battlegroup/SuicideAndNonSuicideUnitTypes.java

```diff
@@ -0,0 +1,122 @@
+ * together, then it would be difficult to figure out if a non-suicideOnHit unit fired the shot or
+ * if a suicideOnHit unit fired the shot. By forcing all of the same unit type to be together when
+ * they are suicideOnHit, then any of the units are fine to commit suicide since they are all the
+ * same.
+ */
+@Value
+class SuicideAndNonSuicideUnitTypes {
+  Collection<UnitType> suicide = new HashSet<>();
+  Collection<UnitType> nonSuicide = new HashSet<>();
+
+  SuicideAndNonSuicideUnitTypes(final UnitType unitType) {
+    addUnitType(unitType);
+  }
+
+  SuicideAndNonSuicideUnitTypes(final Collection<UnitType> unitTypes) {
+    unitTypes.forEach(this::addUnitType);
+  }
+
+  private void addUnitType(final UnitType unitType) {
+    final UnitAttachment unitAttachment = UnitAttachment.get(unitType);
+    if (unitAttachment.getIsSuicideOnHit()) {
+      suicide.add(unitType);
+    } else {
+      nonSuicide.add(unitType);
+    }
+  }
+
+  SuicideAndNonSuicideUnitTypes merge(final SuicideAndNonSuicideUnitTypes other) {
+    suicide.addAll(other.suicide);
+    nonSuicide.addAll(other.nonSuicide);
+    return this;
+  }
+
+  Collection<FiringSquadron> groupFiringSquadrons(final FiringSquadron firingSquadron) {
+    if (size() == 1) {
+      // everything in the firing squadron is either non suicide or suicide of a single unit type
+      // so no need to split it
+      return List.of(firingSquadron);
+    } else if (isSingleSuicideUnitTypeWithNonSuicide()) {
+      return groupSingleSuicideUnitTypeWithNonSuicide(firingSquadron);
+    } else {
+      return groupMultipleSuicide(firingSquadron);
+    }
+  }
+
+  private int size() {
+    return suicide.size() + (nonSuicide.isEmpty() ? 0 : 1);
+  }
+
+  private boolean isSingleSuicideUnitTypeWithNonSuicide() {
+    return size() == 2 && !nonSuicide.isEmpty();
+  }
+
+  private List<FiringSquadron> groupSingleSuicideUnitTypeWithNonSuicide(
+      final FiringSquadron firingSquadron) {
+    return List.of(
+        firingSquadron.toBuilder()
+            .firingUnits(
+                firingSquadron
+                    .getFiringUnits()
+                    .and(
+                        firingUnitFilterData ->
+                            nonSuicide.contains(firingUnitFilterData.getFiringUnit().getType())))
+            .build(),
+        firingSquadron.toBuilder()
+            .name(firingSquadron.getName() + " suicide")
+            .firingUnits(
+                firingSquadron
+                    .getFiringUnits()
+                    .and(
+                        firingUnitFilterData ->
+                            suicide.contains(firingUnitFilterData.getFiringUnit().getType())))
+            .build());
+  }
+
+  private Collection<FiringSquadron> groupMultipleSuicide(final FiringSquadron firingSquadron) {
+    final Collection<FiringSquadron> firingSquadrons =
+        suicide.stream()
+            .map(
```

Comment:
This gets deeply nested. At a glance, this looks to be pretty dense. Would you mind taking a pass over this and similar blocks and trying to make it all easier and faster to understand?

Reply:
It has been removed in https://github.com/triplea-game/triplea/pull/8602/commits/bb56e98c44c6efaec443b627729f2627d646961f